### DISCUSSION TITLE
Equalizer in the web stream with mono output via I2S

### DISCUSSION THREAD

**Discussion by dinggellinggel** [pschatzmann/arduino-audio-tools#1703] [Q&A] [answered]
Is it possible to integrate the equalizer into a webstream? With the following code I was only able to achieve limited success. The stream is played with the equalizer, but it is very slowed down. Thank you very much 

`#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"

URLStream url("xxxx", "xxxx");

I2SStream i2s; 
ConfigEquilizer3Bands cfg_eq; 

VolumeStream volume(i2s);
Equilizer3Bands eq(volume);
EncodedAudioStream dec(&eq, new MP3DecoderHelix()); 
StreamCopy copier(dec, url); 

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  auto config = i2s.defaultConfig(TX_MODE);
  config.pin_ws = 14;
  config.pin_bck = 13;
  config.pin_data = 12;
  config.channels = 1;  
  i2s.begin(config);

  cfg_eq = eq.defaultConfig();
  cfg_eq.setAudioInfo(config); 
  cfg_eq.gain_low = 1.5;  
  cfg_eq.gain_medium = 0.5;
  cfg_eq.gain_high = 0.3; 
  eq.begin(cfg_eq);

  dec.begin();
  volume.begin(config);
  volume.setVolume(0.55); 

  url.begin("Test.mp3", "audio/mp3");
}

void loop() {
  copier.copy();
}`

**Comment by pschatzmann** [comment, marked as answer]
I guess you just forgot to set the log level to a proper value! like described in the
- The most important basics
- It's not working

**Reply by dinggellinggel** [reply]
Thank you for your message and your work!
The code is now running. The stream is playing and the equalizer is also working so far. However, the URL stream is lagging behind the audio output. Sometimes there are dropouts and then the audio output suddenly jumps ahead. What could be the reason for this? (like when a CD used to hang and jump to a different position in the song :D) Hardware: Arduino ESP32 Nano & MAX98357

`#include "AudioTools.h"`
`#include "AudioCodecs/CodecMP3Helix.h"`

`URLStream url("xxxx", "xxxx");`

`I2SStream out;`
`ConfigEquilizer3Bands cfg_eq;`
`MP3DecoderHelix codec;`

`Equilizer3Bands eq(out);`
`EncodedAudioStream dec(&eq, &codec);`
`StreamCopy copier(dec, url);`


`void setup() {`

`auto config = out.defaultConfig(TX_MODE);`
`out.begin(config);`

`cfg_eq = eq.defaultConfig();`
`cfg_eq.gain_low = 0.175; `
`cfg_eq.gain_medium = 0.05;`
`cfg_eq.gain_high = 0.05;`
`cfg_eq.channels = 1;`
`eq.begin(cfg_eq);`

`dec.begin();`

`url.begin("http://icecast.ndr.de/ndr/ndr1niedersachsen/osnabrueck/mp3/128/stream.mp3", "audio/mp3");`
`}`

`void loop() {`
 `copier.copy();`
`}`

**Reply by dinggellinggel** [reply]
Cause found, it was the sample rate.
HelixMP3Encoder = 48000
Equalizer = 44100

Thank you.

**Comment by pschatzmann** [comment]
I must be guessing: Bad Wifi connection, network problems, inproper codec memory settings...
Perform some systematic testing to nerrow down your problem
